### PR TITLE
[Wallet] add option for a custom extended master privat key (xpriv)

### PR DESCRIPTION
**Pull request by jonasschnelli**
At the moment, the HD seed cannot be set by the user. It will always be generated by Bitcoin-Core (`CKey::MakeNewKey()`).

This PR adds a startup argument `-hdxpriv` where users can set their custom hd extended master private key. The argument takes only affect during the creation of a new `wallet.dat` and there will be a warning if the argument is present during loading an exiting wallet.

If a custom extended master private key has been set, it will be kept when encrypting the wallet (with log-printing a warning).

### CODE REVIEW

**Comment by btcdrak** [comment]
As it stands this will leak the xpriv since the xpriv will be in the command history. The import of the xpriv should be interactive and should be encrypted before being stored.

I know previously we would not recommend forcing encryption because it could lead to funds loss, here the user already has the xpriv so it's already backed up.

**Comment by paveljanik** [comment]
@btcdrak We can change this to accept the seed only from stdin (`-stdin`).

**Comment by jonasschnelli** [comment]
Yes. I'm aware of the security weakness of accepting a xpriv over a startup argument.
A previous proposed solution would be to support a tool called `bitcoin-wallet` which could create, alter and encrypt wallet without the need of bitcoind (this would by my preferred solution but can be added later).

I think accepting a xpriv over a configuration value is not extremely bad. If an attacker can access bitcoin.conf he might also be capable of grabbing your RPC password and dump your wallet to /tmp.

**Comment by btcdrak** [comment]
Personally I prefer a separate interactive import tool and a Qt GUI option which I think would be used more in any case.

**Comment by btcdrak** [comment]
> @btcdrak We can change this to accept the seed only from stdin (-stdin).

Sounds good, that relieves my concerns.

File: src/wallet/walletdb.h [hunk 1]
```diff
@@ -47,8 +47,12 @@ class CHDChain
 public:
     uint32_t nExternalChainCounter;
     CKeyID masterKeyID; //!< master key hash160
+    CExtKey customMasterKey;
+
+    static const int VERSION_BASE = 1;
+    static const int VERSION_WITH_CUSTOM_MASTER_KEY = 2;
+    static const int CURRENT_VERSION = VERSION_WITH_CUSTOM_MASTER_KEY;
```
Comment (TheBlueMatt) [inline src/wallet/walletdb.h]: Why add a new version? Why not just overwrite more akin to the way #11085 works?

**Review by TheBlueMatt — COMMENTED** [review]
Needs rebase.

**Comment by sipa** [comment]
Concept ACK, but I think this needs significant reworking to interact with multiwallet.

**Comment by jnewbery** [comment]
This would possibly be easier to implement using bitcoin-wallet-tool (https://github.com/bitcoin/bitcoin/pull/8745)

**Comment by nerdDan** [comment]
I am confused. Is there any workaround that allows me importing xpriv into `bitcoind`?

**Comment by nerdDan** [comment]
I am confused. Is there any workaround that allows me importing a xpriv?

**Pull request closed without merging**